echo_test_server.py

```diff
@@ -0,0 +1,105 @@
+import asyncio
+import os
+import signal
+import sys
+from textwrap import dedent
+
+try:
+    from rich import print
+except ImportError:
+    pass
+
+os.environ["RUST_LOG"] = "mitmproxy_wireguard=debug"
+
+import mitmproxy_wireguard
+
+# (private key, public key)
+server_keypair = ("EG47ZWjYjr+Y97TQ1A7sVl7Xn3mMWDnvjU/VxU769ls=", "mitmV5Wo7pRJrHNAKhZEI0nzqqeO8u4fXG+zUbZEXA0=")
+client_keypair = ("qG8b7LI/s+ezngWpXqj5A7Nj988hbGL+eQ8ePki0iHk=", "Test1sbpTFmJULgSlJ5hJ1RdzsXWrl3Mg7k9UTN//jE=")
+
+
+async def main():
+    print(f"{dir(mitmproxy_wireguard)=}")
+
+    k = mitmproxy_wireguard.genkey()
+    print(f"mitmproxy_wireguard.genkey()={k!r}")
+    print(f"{mitmproxy_wireguard.pubkey(k)=}")
+
+    def receive_datagram(data, src_addr, dst_addr):
+        print(f"Received datagram: {data=} {src_addr=} {dst_addr=}")
+        server.send_datagram(data.upper(), dst_addr, src_addr)
+        print("Echoed datagram.")
+
+    server = await mitmproxy_wireguard.start_server(
+        "0.0.0.0",
+        51820,
+        server_keypair[0],
+        [client_keypair[1]],
+        handle_connection,
+        receive_datagram
+    )
+    print(f"{server.getsockname()=}")
+
+    print(dedent(f"""
+    :white_check_mark: Server started. Use the following WireGuard config for testing:
+    ------------------------------------------------------------
+    [Interface]
+    PrivateKey = {client_keypair[0]}
+    Address = 10.0.0.1/32
+    
+    [Peer]
+    PublicKey = {server_keypair[1]}
+    AllowedIPs = 10.0.0.0/24
+    Endpoint = 127.0.0.1:51820
+    ------------------------------------------------------------
+    
+    And then run `nc 10.0.0.42 1234` or `nc -u 10.0.0.42 1234` to talk to the echo server.
+    """))
+
+    def stop(*_):
+        print("Stopping server...")
+        server.stop()
+        print("Stopped.")
+        signal.signal(signal.SIGINT, lambda *_: sys.exit())
+
+    signal.signal(signal.SIGINT, stop)
+    await asyncio.sleep(9999)
+
+
+async def handle_connection(r: asyncio.StreamReader, w: asyncio.StreamWriter):
```

Comment:
TCP connections get a StreamReader and a StreamWriter. In our case that's exactly the same object, but this split mirrors what `asyncio.start_server` is doing. I'm open to change that and combine it later on. :)

Ideally we'd also fully type out our Rust Stream implementation instead of pretending that it's an `asyncio.StreamReader`.